### PR TITLE
feature(category) Adding category tree autcomplete

### PR DESCRIPTION
**Pull request by wiewiurdp**
# Description

Category tree autocomplete endpoint

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Unit Tests
- [x] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php [hunk 1]
```diff
@@ -70,6 +70,35 @@ public function findTreeIdByCode(string $code): ?CategoryTreeId
...
+        string $field = null,
+        ?string $order = 'ASC'
+    ): array {
+        $query = $this->connection->createQueryBuilder()
+            ->select('id, code, COALESCE(name->>:language, null) as label')
+            ->from(self::TREE_TABLE, 'c')
+            ->setParameter(':language', $language->getCode());
+
+        if ($search) {
+            $query->orWhere('code ILIKE :search');
```
Comment (wfajczyk) [inline module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php:86]: I think is beter.
```suggestion
            $query->andWhere('code ILIKE :search');
```

**Review by wfajczyk — CHANGES_REQUESTED** [review]

File: module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php (hunk 1 above)
Comment (piotrkreft) [reply module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php:86]: just where I guess. There's no condition at this point

File: module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php [hunk 2]
```diff
@@ -70,6 +70,35 @@ public function findTreeIdByCode(string $code): ?CategoryTreeId
...
 
+    public function autocomplete(
+        Language $language,
+        string $search = null,
+        int $limit = null,
+        string $field = null,
+        ?string $order = 'ASC'
+    ): array {
+        $query = $this->connection->createQueryBuilder()
+            ->select('id, code, COALESCE(name->>:language, null) as label')
```
Comment (piotrkreft) [inline module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php:81]: COALESCE is not needed probably?
```suggestion
            ->select('id, code, name->>:language as label')
```

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php (hunk 2 above)
Comment (wiewiurdp) [reply module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php:81]: I think it is need as name is not required  and can be empty an array.

Comment (piotrkreft) [reply module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php:81]: Basically what you're doing with COALESCE is saying if the value is null return null... So I think it's not ;)

File: module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php [hunk 3]
```diff
@@ -70,6 +70,35 @@ public function findTreeIdByCode(string $code): ?CategoryTreeId
...
+            ->from(self::TREE_TABLE, 'c')
+            ->setParameter(':language', $language->getCode());
+
+        if ($search) {
+            $query->Where('code ILIKE :search');
```
Comment (piotrkreft) [inline module/category/src/Infrastructure/Persistence/Query/DbalTreeQuery.php:86]: should be lower case. I though it is cover in one of the checks but apparently wrong

**Review by piotrkreft — APPROVED** [review]

**Review by wfajczyk — APPROVED** [review]

**Pull request merged**